{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tiled Primer Design\n",
    "\n",
    "This jupyter notebook semi-automates the process of designing primers using k-mers for tiled PCR amplification across random sites in a single genome. \n",
    "\n",
    "\n",
    "### Credit\n",
    "\n",
    "Written by Julie Chen and Gowtham Thakku"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The necessary python packages are imported here. If you receive an error, you may have to install these packages\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Blast import NCBIXML\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools\n",
    "import re\n",
    "import xlrd\n",
    "import random\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declare some modifiable parameters here:\n",
    "\n",
    "date = '20220713'\n",
    "kmerSize = 5\n",
    "genomeFileName = '../NC_000962.fasta'\n",
    "record = SeqIO.read(genomeFileName, 'fasta')\n",
    "fastaSeq = record.seq\n",
    "\n",
    "# note: few more manual inputs later (see Doc for more instructions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Designing Primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract initial primer sequences using a chosen k-mer\n",
    "\n",
    "**How to tell if a k-mer will likely self-dimerize**: Does the reverse (NOT rc) mostly bind to the forward sequence of the k-mer? If so, don't use this k-mer. \\\n",
    "e.g. F: 5' GCTGC 3' would be R: 3' CGTCG 5' which is complementary at 4/5 positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# based on the \"compatible\" k-mers identified, select a set of k-mers to use\n",
    "df_kmers = pd.read_csv('kmers_diverse_ranked.csv')\n",
    "df_kmers = df_kmers[df_kmers['num. of occurrences'] >= 1] \n",
    "\n",
    "\n",
    "# Specify any sequences here that you do not want the chosen kmers to be complementary to\n",
    "# (e.g. if you already have primer sequences to include in the pool, list them here so any new kmers don't bind to them)\n",
    "excludeSeq = ['TCTTTAAAATCGCGTTCGCCCTTCGCAATTCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'TCTTTCAGGTCGAGTACGCCTTCTTGTTGGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'TCTTTGGCCATGATCGACACTTGCGACTTGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'AGTACCCGATCATATTGGGCAACAACTGATCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'GCTGCCCACTCCGAATCGTGCTGACCGCGGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'GCTGCTCGGCGTCGATAAGATGAGAAGAGGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'GCTGCGTTCGCGGTAGCCCGCCCCGCACAGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'GCTGCCATCAGATTGGCTGCGTAGTGGGTTCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'GCTGCTCGGGCTCGGGTGAGGACCTCGAGGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'ACACCCCACCGCAAACCATCGGGGCCCTCGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'ACACCGTAGTTGGCGGCGTGGACGCGGCTGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'CGGCCCCGTCCTCGGCGGAGGTGACCTGGACCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'CATCGACCTGCGCCTGGCGCACCCACTTACCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'CATCGTGGAAGCGACCCGCCAGCCCAGGATCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'TGAGCACCAGGGCGTCGGCGGCGAGGAAGGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'GAGCTTCCGACCGCTCCGACCGACGGTTGGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'TTGGTCATCAGCCGTTCGACGGTGCATCTGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'ACCACCTGACATGACCCCATCCTTTCCAAGCCCTATAGTGAGTCGTATTAATTTC',\\\n",
    "'CGGCCAGCACGCTAATTACCCGGTTCATCG',\\\n",
    "'AGTACACATCGATCCGGTTCAGCGAGCGGC',\\\n",
    "'CGGCCCGTATACCTTCCTCGCCGCCGACGC',\\\n",
    "'ACACCGCCCGCACCGACCTGCTGGCGTTCA',\\\n",
    "'GCTGCGCGGAGACGGTGCGTAAGTGGGTGC',\\\n",
    "'GCTGCGCGGGCTGCTCTCGACGTTCATCGC',\\\n",
    "'ACACCCGTGCCGCAACCATCGACGTCGCGA',\\\n",
    "'GCTGCACTCCATCTACGACCAGCCCGACGC',\\\n",
    "'ACACCCCAGCACTGACCACCTAGACTGCCA',\\\n",
    "'CATCGACCTACTACGACCACATCAACCGGG',\\\n",
    "'CATCGAGGTGGCCAGATGCACCGTCGAACG',\\\n",
    "'CGGCCTGTCCGGGGTCGCGCTGGTCACCAG',\\\n",
    "'TGAGCTGAAGCGCTTGCGGCGGGACAACGC',\\\n",
    "'AAGCCATCTGGACCCGCCAACAAGAAGGCG',\\\n",
    "'ACACCTTGATCGCCACCGGCGTCAACGCCG',\\\n",
    "'CGGCCTATACAAGACCGAGCTGATCAAACC',\\\n",
    "'AAGCCCGCAGGACCACGATCGCTGATCCGG',\\\n",
    "'TGAGCGGGCGGTGCGGATGGTCGCAGAGAT']\n",
    "\n",
    "for currSeq in excludeSeq:\n",
    "    for i in range(len(currSeq)-5):\n",
    "        curr5mer = currSeq[i:i+5]\n",
    "        df_kmers.drop(df_kmers[df_kmers['k-mer sequence']==curr5mer].index, inplace=True)\n",
    "  \n",
    "\n",
    "chosenKmers = list(df_kmers['k-mer sequence'])\n",
    "\n",
    "## If you want to manually specify the list of possible kmers, do so here\n",
    "# chosenKmers = ['GCATC', 'ATCCG', 'CAGGT', 'TCCAG', 'CGCAT', 'GCTCA', \\\n",
    "#                'CGACT', 'ATCAG', 'TCCGA', 'GACAT', 'TCAGG', 'GCACT', \\\n",
    "#                'GCAAT', 'TCAAG', 'CGCTA', 'GACTC', 'CTCAG', 'GCTAT', \\\n",
    "#                'GACTA', 'GCTAA', 'CTAAG', 'GTTCA', 'GCTAC', 'GCCAT']\n",
    "\n",
    "#PARAMETERS CAN BE CHANGED\n",
    "#Set other parameters for run\n",
    "gRNA_size = 28 # This is the length of the crRNA spacer sequence\n",
    "wiggleRoom = 20 # This determines the range of possible distances between forward and reverse primers\n",
    "primerLength = 30 # This is the length of the primers to be designed (can be truncated later if needed)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to find all positions of a particular kmer in the genome\n",
    "def listAllPositions(kmer, genomeSeq = fastaSeq):\n",
    "\n",
    "    pattern = re.compile(r'(?=(' + kmer + '))')\n",
    "    positions = [p.start() for p in re.finditer(pattern, str(genomeSeq))]\n",
    "\n",
    "    return positions\n",
    "\n",
    "# function to find the reverse complement of a given sequence\n",
    "def revComplementSeq(seqtoRC):\n",
    "\n",
    "    seqObject = Seq(seqtoRC)\n",
    "    rc = seqObject.reverse_complement()\n",
    "\n",
    "    return rc\n",
    "\n",
    "# function to find 3'-end positions for compatible forward kmer positions\n",
    "def find3EndF(kmerIndices):\n",
    "\n",
    "    ends = [(i + kmerSize) for i in kmerIndices]\n",
    "\n",
    "    return ends\n",
    "\n",
    "\n",
    "def gcContentFrac(seqGC):\n",
    "    gcNum = seqGC.count('G')+seqGC.count('g')+seqGC.count('C')+seqGC.count('c')\n",
    "    gcFrac = gcNum/len(seqGC)\n",
    "    \n",
    "    return gcFrac\n",
    "\n",
    "\n",
    "#function to find distance between 3' ends of foward and reverse primers\n",
    "def positionDifference(revList, forList):\n",
    "\n",
    "    differences = [(revList[r] - forList[r]) for r in range(len(revList))]\n",
    "    return differences\n",
    "\n",
    "#function to get forward primer sequences\n",
    "def initialPrimerSeqF(positionsF, genomeSeq = fastaSeq, length = primerLength-kmerSize):\n",
    "\n",
    "    primerSequences = []\n",
    "\n",
    "    for i in positionsF:\n",
    "\n",
    "        primerStart = i - length\n",
    "        primerEnd = i + kmerSize\n",
    "        primerSequences.append(str(genomeSeq[primerStart:primerEnd]))\n",
    "\n",
    "    return primerSequences\n",
    "\n",
    "\n",
    "#function to get reverse primer sequences\n",
    "def initialPrimerSeqR(positionsR, genomeSeq = fastaSeq, length = primerLength-kmerSize):\n",
    "\n",
    "    primerSequences = []\n",
    "\n",
    "    for i in positionsR:\n",
    "\n",
    "        primerEnd = i + kmerSize + length\n",
    "        rcSeq = revComplementSeq(str(genomeSeq[i:primerEnd]))\n",
    "        primerSequences.append(str(rcSeq))\n",
    "\n",
    "    return primerSequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#function to generate all possible compatible primers for a given set of k-mers\n",
    "\n",
    "def numCompatiblePrimers(chosenKmers, gRNA_size, wiggleRoom, primerLength, genomeSeq = fastaSeq):\n",
    "    \n",
    "    kmerPositionsF = []\n",
    "\n",
    "    kmerSeqColumn = 'k-mer sequence'\n",
    "\n",
    "    # loop to identify all compatible positions of kmers within acceptable distance\n",
    "\n",
    "    for currKmer in chosenKmers:\n",
    "\n",
    "\n",
    "        positions_currKmer = listAllPositions(currKmer, genomeSeq)\n",
    "        positions_currKmer = [t for t in positions_currKmer if t >50 and t <2878800]\n",
    "        compatiblePositions = []\n",
    "\n",
    "        for currPos in positions_currKmer:\n",
    "\n",
    "            primerSpace_start = currPos + kmerSize + gRNA_size \n",
    "            primerSpace_end = primerSpace_start + wiggleRoom + kmerSize\n",
    "\n",
    "            if primerSpace_end < len(genomeSeq):\n",
    "                searchSpace = str(genomeSeq[primerSpace_start:primerSpace_end])\n",
    "\n",
    "            for searchKmer in chosenKmers:\n",
    "                kmer_rc = revComplementSeq(searchKmer) \n",
    "                if searchSpace.find(str(kmer_rc)) > -1:\n",
    "                    compatiblePositions.append(currPos)\n",
    "                    break\n",
    "\n",
    "        kmerPositionsF.extend(compatiblePositions)\n",
    "\n",
    "#     print(\"Number of F primer Positions: \",len(kmerPositionsF))\n",
    "\n",
    "    positions3F = find3EndF(kmerPositionsF)\n",
    "\n",
    "\n",
    "    # function to find 3'-end positions for compatible reverse kmer positions\n",
    "    kmerPositionsR = []\n",
    "\n",
    "    kmerSeqColumn = 'k-mer sequence'\n",
    "\n",
    "    revPositions = []\n",
    "    positions_rc = []\n",
    "    for searchKmer in chosenKmers:     \n",
    "        kmer_rc = revComplementSeq(searchKmer)\n",
    "        positions_rc.extend(listAllPositions(str(kmer_rc), genomeSeq))\n",
    "\n",
    "    for i in positions3F:\n",
    "\n",
    "\n",
    "        primerSpace_start = i + gRNA_size \n",
    "        primerSpace_end = primerSpace_start + wiggleRoom + kmerSize\n",
    "\n",
    "        for j in positions_rc:\n",
    "            if primerSpace_start <= j <= primerSpace_end:\n",
    "                revPositions.append(j)\n",
    "                break\n",
    "\n",
    "    kmerPositionsR.extend(revPositions)\n",
    "    \n",
    "    if len(kmerPositionsR) < len(kmerPositionsF):\n",
    "#         print(\"F and R are different!\")\n",
    "        kmerPositionsF = kmerPositionsF[:-1]\n",
    "        positions3F = positions3F[:-1]\n",
    "\n",
    "#     print(\"Number of R primer Positions: \",len(kmerPositionsR))\n",
    "\n",
    "\n",
    "    kmerPositionDiff = positionDifference(kmerPositionsR, positions3F)\n",
    "\n",
    "\n",
    "\n",
    "    primerSeqF = initialPrimerSeqF(kmerPositionsF)\n",
    "    primerSeqR = initialPrimerSeqR(kmerPositionsR)\n",
    "    \n",
    "    gcFractionR = []\n",
    "    gcFractionF = []\n",
    "    for currPrimer in primerSeqF:\n",
    "        gcFractionF.append(gcContentFrac(currPrimer))\n",
    "    for currPrimer in primerSeqR:\n",
    "        gcFractionR.append(gcContentFrac(currPrimer))\n",
    "\n",
    "    primerInitialSeq = {'primerF sequence': primerSeqF, 'primerR sequence': primerSeqR,\n",
    "                        'kmer index': kmerPositionsF,\n",
    "                        'primerF 3': positions3F, 'primerR 3': kmerPositionsR, \n",
    "                        'position difference': kmerPositionDiff, 'gcFracF': gcFractionF, 'gcFracR': gcFractionR}\n",
    "\n",
    "    if len(primerInitialSeq) == 0:\n",
    "        return pd.DataFrame()\n",
    "    \n",
    "    df_primerSeq = pd.DataFrame(primerInitialSeq)\n",
    "    df_primerSeq.head()\n",
    "\n",
    "    df_primerSeq_filtered = df_primerSeq\n",
    "\n",
    "\n",
    "    #Adapter to add to the forward primers\n",
    "    adaptPrimerF = 'gaaatTAATACGACTCACTATAGGG'\n",
    "    # adaptGuide = 'GTTTTAGTCCCCTTCGTTTTTGGGGTAGTCTAAATCCCCTATAGTGAGTCGTATTAatttc'\n",
    "\n",
    "#     df_primerSeq_filtered['primerF sequence'] = adaptPrimerF + df_primerSeq_filtered[\"primerF sequence\"].astype(str)\n",
    "\n",
    "    fastaSeqOccupied = np.zeros(len(fastaSeq), dtype=int)\n",
    "#     print(\"Filtering sequence:\", chosenKmers)\n",
    "    \n",
    "    #Screening primers to see if kmer binding sites exist\n",
    "    for currKmer in chosenKmers:\n",
    "#         \n",
    "        kmer_rc = str(revComplementSeq(currKmer))\n",
    "        df_primerSeq_filtered[\"Fexclude\"]=df_primerSeq_filtered[\"primerF sequence\"].str.find(kmer_rc)\n",
    "        df_primerSeq_filtered[\"Rexclude\"]=df_primerSeq_filtered[\"primerR sequence\"].str.find(kmer_rc)\n",
    "        df_primerSeq_filtered = df_primerSeq_filtered[df_primerSeq_filtered.Fexclude < 0]\n",
    "        df_primerSeq_filtered = df_primerSeq_filtered[df_primerSeq_filtered.Rexclude < 0]\n",
    "        \n",
    "      \n",
    "    if df_primerSeq_filtered.shape[0] == 0:\n",
    "        return pd.DataFrame()\n",
    "    \n",
    "    for index, row in df_primerSeq_filtered.iterrows():\n",
    "        currF = row['primerF 3']\n",
    "        currR = row['primerR 3']\n",
    "\n",
    "        df_primerSeq_filtered.loc[index,'ampUnique'] = sum(fastaSeqUnique[currF:currR-28])\n",
    "        df_primerSeq_filtered.loc[index,'fUnique10'] = sum(fastaSeqUnique[currF-28-5:currF-28])\n",
    "        df_primerSeq_filtered.loc[index,'rUnique10'] = sum(fastaSeqUnique[currR:currR+5])\n",
    "        df_primerSeq_filtered.loc[index,'amplicon'] = fastaSeq[currF:currR]\n",
    "        if sum(fastaSeqOccupied[currF-20:currR+20]>0):\n",
    "            df_primerSeq_filtered.loc[index,'occupied'] = 2\n",
    "        else:\n",
    "            df_primerSeq_filtered.loc[index,'occupied'] = 1\n",
    "            fastaSeqOccupied[currF-20:currR+20] = np.ones(len(range(currF-20,currR+20)))\n",
    "\n",
    "        primerF = row['primerF sequence']\n",
    "        primerR = row['primerR sequence']\n",
    "    \n",
    "    \n",
    "    df_primerSeq_filtered = df_primerSeq_filtered[df_primerSeq_filtered.gcFracF<0.7]\n",
    "    df_primerSeq_filtered = df_primerSeq_filtered[df_primerSeq_filtered.gcFracR<0.7]    \n",
    "    df_primerSeq_filtered = df_primerSeq_filtered[df_primerSeq_filtered.ampUnique>0]\n",
    "    df_primerSeq_filtered = df_primerSeq_filtered[df_primerSeq_filtered.fUnique10>0]\n",
    "    df_primerSeq_filtered = df_primerSeq_filtered[df_primerSeq_filtered.rUnique10>0]\n",
    "    df_primerSeq_filtered = df_primerSeq_filtered[df_primerSeq_filtered.occupied<2]\n",
    "    \n",
    "#     print(\"Remaining primer pairs:\", df_primerSeq_filtered.shape)\n",
    "        \n",
    "    return df_primerSeq_filtered\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this snippet to expand an existing primer pool (made with pre-defined kmers) by one additional k-mer\n",
    "\n",
    "numBootStrap = 10 #This value determines the number of random searches done at each pool size\n",
    "\n",
    "\n",
    "numPrimers = [] \n",
    "numKmers = 1 #Number of new kmers added at each step of pool expansion\n",
    "\n",
    "for currNum in range(numBootStrap):\n",
    "    kmerBootStrap = ['AAAGA','ACCAA','AGCTC','CGATG','GCAGC','GCTCA','GGCCG','GGCTT','GGTGT','GTACT','GTGGT']\n",
    "    bootSample = random.sample(range(0,len(chosenKmers)), numKmers)\n",
    "    for i in bootSample:\n",
    "        kmerBootStrap.append(chosenKmers[int(i)])\n",
    "    try:\n",
    "        df_boot = numCompatiblePrimers(kmerBootStrap, gRNA_size, wiggleRoom, primerLength, genomeSeq=fastaSeq) \n",
    "#         except:\n",
    "#             df_boot = pd.DataFrame()\n",
    "        if df_boot.shape[0]>0:\n",
    "            print(df_boot.shape[0])\n",
    "            print(kmerBootStrap)\n",
    "        numPrimers.append(df_boot.shape[0])\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "# kmerStarterDict.loc[x,'numPrimers'] = numPrimers\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NumKmers\tPoolSize\n",
      "1 \t\t 2\n",
      "1 \t\t 10\n",
      "1 \t\t 13\n",
      "1 \t\t 21\n",
      "1 \t\t 31\n",
      "2 \t\t 134\n",
      "2 \t\t 160\n",
      "3 \t\t 208\n",
      "3 \t\t 403\n",
      "4 \t\t 510\n",
      "4 \t\t 517\n",
      "4 \t\t 554\n",
      "4 \t\t 570\n",
      "5 \t\t 638\n",
      "5 \t\t 700\n",
      "5 \t\t 729\n",
      "5 \t\t 757\n",
      "6 \t\t 802\n",
      "6 \t\t 865\n",
      "6 \t\t 935\n",
      "List of kmers: ['GTTGT', 'TGACG', 'TCGGT', 'CACGG', 'ATGTC', 'CGGAA']\n"
     ]
    }
   ],
   "source": [
    "# Run this snippet to generate a new primer pool without any pre-defined kmers\n",
    "\n",
    "numBootStrap = 10 #This value determines the number of random searches done at each kmer pool size \n",
    "maxNumKmers = 6 #This value determines the maximum kmer pool size searched\n",
    "\n",
    "fastaSeqUnique = np.ones(len(fastaSeq,), dtype=int)\n",
    "numPrimers = []\n",
    "max_shape = 0;\n",
    "kmerBoot = []\n",
    "best_kmer = []\n",
    "print(\"NumKmers\\tPoolSize\")\n",
    "for numKmers in [1]*maxNumKmers:\n",
    "    \n",
    "    for currNum in range(numBootStrap):\n",
    "        kmerBootStrap = kmerBoot.copy()\n",
    "        bootSample = random.sample(range(0,len(chosenKmers)), numKmers)\n",
    "        for i in bootSample:\n",
    "            kmerBootStrap.append(chosenKmers[int(i)])\n",
    "        try:\n",
    "            df_boot = numCompatiblePrimers(kmerBootStrap, gRNA_size, wiggleRoom, primerLength, genomeSeq=fastaSeq)          \n",
    "            if df_boot.shape[0]>max_shape:\n",
    "                print(len(kmerBootStrap),\"\\t\\t\", df_boot.shape[0])\n",
    "                best_kmer = kmerBootStrap.copy()              \n",
    "                numPrimers.append(df_boot.shape[0])\n",
    "                max_shape = max(numPrimers)\n",
    "        except:\n",
    "            df_boot = pd.DataFrame()\n",
    "    kmerBoot = best_kmer.copy()\n",
    "# kmerStarterDict.loc[x,'numPrimers'] = numPrimers\n",
    "print(\"List of kmers:\", kmerBoot)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
